llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp

```diff
@@ -640,25 +640,31 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
                                                     DemandedMask, Known))
             return R;
 
-      // TODO: If we only want bits that already match the signbit then we don't
+      uint64_t ShiftAmt = SA->getLimitedValue(BitWidth - 1);
+      // If we only want bits that already match the signbit then we don't
       // need to shift.
+      if (DemandedMask.countr_zero() >= ShiftAmt) {
+        unsigned NumLowDemandedBits = BitWidth - DemandedMask.countr_zero();
+        unsigned SignBits =
+            ComputeNumSignBits(I->getOperand(0), Depth + 1, CxtI);
+        if (SignBits > ShiftAmt && SignBits - ShiftAmt >= NumLowDemandedBits)
```

Comment:
This can be `>=`:
https://alive2.llvm.org/ce/z/C3VNoR

Reply:
@goldsteinn maybe your link is wrong. 
anyway. considering that if signbits equal shiftamt, 
https://alive2.llvm.org/ce/z/82A769
It doesn't seem to work.

Reply:
I think that @goldsteinn is in principle correct, but if `SignBits == ShiftAmt` then `SignBits - ShiftAmt` is 0, so the condition would become `0 >= NumHighDemandedBits`, which is the same as `NumHighDemandedBits == 0`, which is a degenerate case that gets handled earlier anyway. So this just doesn't matter either way.

Reply:
I'm update alive link for showing verification of case `SignBits == ShiftAmt` and `SignBits > ShiftAmt && SignBits - ShiftAmt >= NumHighDemandedBits`.  if I misunderstand of it, please let me know.